Add docs and guide for MLflow Gateway (#8763)
mlflow/mlflow: docs/source/gateway.rst

.. _gateway:

=============================
MLflow Gateway (Experimental)
=============================

.. warning::

    The MLflow Gateway is a new, **experimental feature**. It is subject to modification, feature improvements, or feature removal without advance notice.

> The `MLflow Gateway` service is a powerful tool designed to streamline the usage and management of

Reviewer: I think we should remove backticks here

> various language model providers, such as `OpenAI` and `Anthropic`, within an organization.

Reviewer: Nit: I think we should remove the backticks around vendor names too

It offers a high-level interface that simplifies the interaction with these services by providing
a unified endpoint to handle all language model requests.

A major advantage of using the `MLflow Gateway` service is its centralized management of API keys.
By storing these keys in one secure location, organizations can significantly enhance their

> security posture by minimizing the exposure of sensitive API keys throughout the system. It also
> helps to prevent exposing these keys within code or having end-users have to manage keys safely.

Reviewer: ```suggestion
security posture by minimizing the exposure of sensitive API keys throughout the system. It also
helps to prevent exposing these keys within code or requiring end-users to manage keys safely.
```

The gateway is designed to be flexible and adaptable, capable of hot-swapping routes if the

> configuration file at the target location is overwritten. This enables the easy incorporation

Reviewer: I think we should avoid the "hot-swapping" call-out, since we don't want to encourage users to mutate routes that are used in production. On Databricks, we actively prevent users from doing this. We should say that users can easily add and remove routes by updating the configuration file.

of new providers or provider model types into the system without necessitating changes to
applications that interface with the gateway. This level of adaptability makes the `MLflow Gateway`
Service an invaluable tool in environments that require agility and quick response to changes.

This simplification and centralization of language model interactions, coupled with the added
layer of security for API key management, make the `MLflow Gateway` service an ideal choice for

> businesses that utilize language models on a regular basis.

Reviewer: ```suggestion
organizations that use language models on a regular basis.
```

.. contents:: Table of Contents
  :local:
  :depth: 2

.. _gateway-concepts:

Concepts
========

> Gateway Configuration

Reviewer: Can we add a `.. note` in this section indicating that the Databricks has a separate configuration API?

users can access.

The configuration file includes a series of sections, each representing a unique route.
Each route section has a name, a type, and a model specification, which includes the model
provider, name, and configuration details. The configuration section typically contains the base
URL for the API and an environment variable for the API key.

Here is an example of a simple single-route configuration:

.. code-block:: yaml

    routes:
      - name: chat
        type: llm/v1/chat
        model:
          provider: openai
          name: gpt-3.5-turbo
          config:
            openai_api_base: https://api.openai.com/v1
            openai_api_key: $OPENAI_API_KEY

In this example, we define a route named ``chat`` that corresponds to the ``llm/v1/chat`` type, which
will use the ``gpt-3.5-turbo`` model from `OpenAI` to return query responses from the `OpenAI` service.

The beauty of the Gateway Configuration lies in its flexibility and ease of use.
If the configuration file is updated at the target location, the `Mlflow Gateway` service can hot-swap
routes, allowing new providers or model types to be incorporated without requiring changes to

> the applications interfacing with the gateway. The hot-swap functionality is autonomous and

Reviewer: +1 for automatic

> the applications interfacing with the gateway. The hot-swap functionality is autonomous and

Reviewer: I also don't think we should be using the "hot-swap" terminology. I think it's best to tell users that they can easily update the configuration to add or remove routes without any downtime.

requires no server downtime or manual restarting of the service.
This allows for simple, seamless changes to the underlying models and providers while keeping
your applications steady and reliable.

In order to define an API KEY for a given provider, there are three primary options:

1. Directly include it in the YAML configuration file.
2. Use an environment variable to store the API key and reference it in the YAML configuration file.
3. Define your API key in a file and reference the location of that key-bearing file within the YAML configuration file.

If you choose to include the API key directly, replace ``$OPENAI_API_KEY`` in the YAML file with your
actual API key.

If you prefer to use an environment variable (recommended), you can define it in your shell
environment. For example:

.. code-block:: bash

     export OPENAI_API_KEY="your_openai_api_key"

**Note:** Replace "your_openai_api_key" with your actual `OpenAI` API key.

Remember to ensure secure access to the system that the `MLflow Gateway` service is running in to

> protect access to these keys.

Reviewer: Can we call out a specific recommendation, such as running the gateway behind a reverse proxy authentication layer?

Routes
------

`Routes` are central to how the `MLflow Gateway` functions. Each route acts as a proxy endpoint for the

> user, channeling requests to the underlying model and provider specified in the configuration file.

Reviewer: ```suggestion
user, forwarding requests to the underlying `model` and `provider` specified in the configuration file.
```

A route in the `MLflow Gateway` consists of the following fields:

* **name**: This is the unique identifier for the route. This will be part of the URL when making API calls via the `MLflow Gateway`.

> 
> * **type**: The type of the route corresponds to the type of language model interaction you desire. For instance, 'llm/v1/chat' for chat operations, 'llm/v1/completions' for text completion operations, and 'llm/v1/embeddings' for text embeddings.

Reviewer: Can we call out completions and embeddings first, since they're more broadly supported? We only have Chat for OpenAI

```suggestion

* **type**: The type of the route corresponds to the type of language model interaction you desire. For instance, ``llm/v1/completions`` for text completion operations, ``llm/v1/embeddings`` for text embeddings, ``llm/v1/chat`` for chat operations.
```

> * **model**: This encapsulates the details of the model to which this route will direct its requests. The model contains the following details:

Reviewer: ```suggestion
* **model**: Defines the model to which this route will forward requests. The model contains the following details:
```

>     * **provider**: Specifies the name of the ``provider`` for this model. For example, ``openai`` for `OpenAI`'s ``GPT-3`` models.

Reviewer: Can we link to the `Providers` section below, which is awesomely written and does a great job of explaining what a provider is?

Duplicate route names will raise an Exception.

Providers
---------
The `MLflow Gateway` is designed to support a variety of model providers.
A provider represents the source of the machine learning models, such as `OpenAI`, `Anthropic`, and so on.
Each provider has its specific characteristics and configurations that are encapsulated within the model part of a route in the `MLflow Gateway`.

Within each model block in the configuration file, the provider field is used to specify the name
of the provider for that model. This is a string value that needs to correspond to a provider the `MLflow Gateway` supports.

Here's an example of a provider configuration within a route:

.. code-block:: yaml

    routes:
        - name: chat
          type: llm/v1/chat
          model:
            provider: openai
            name: gpt-4
            config:
              openai_api_base: https://api.openai.com/v1
              openai_api_key: $OPENAI_API_KEY

In the above configuration, ``openai`` is the `provider` for the model.

As of now, the `MLflow Gateway` supports the following providers:

* **openai**: This is used for models offered by `OpenAI <https://platform.openai.com/>`_.

> * **anthropic**: This is used for models offered by `Anthropic <https://docs.anthropic.com/claude/docs>`_.

Reviewer: Can we document which route types and models are supported by each provider? Can we create a table for this that we can update over time?

More providers are being added continually. Check the latest version of the `MLflow Gateway` for the
most up-to-date list of supported providers.

Remember, the provider you specify must be one that the `MLflow Gateway` supports. If the provider
is not supported, the Gateway will return an error when trying to route requests to that provider.

> Model Names

Reviewer: I think this section should be called `Models` and should be moved above the section called `Providers`. I think `Providers` should be a subsection of `Models`.

conversational models, while `llm/v1/embeddings` endpoints would typically be associated with
embedding or transformer models. The model you choose should be appropriate for the type of endpoint specified.

Here's an example of a model name configuration within a route:

.. code-block:: yaml

    routes:
      - name: embeddings
        type: llm/v1/embeddings
        model:
          provider: openai
          name: text-embedding-ada-002
          config:
            openai_api_base: https://api.openai.com/v1
            openai_api_key: $OPENAI_API_KEY

In the above configuration, ``text-embedding-ada-002`` is the model used for the embeddings endpoint.

When specifying a model, it is critical that the provider supports the model you are requesting.
For instance, ``openai`` as a provider supports models like ``text-embedding-ada-002``, but other providers
may not. If the model is not supported by the provider, the `MLflow Gateway` will return an error
when trying to route requests to that model.

.. important::

    Always check the latest documentation of the specified provider to ensure that the model you want

>     to use is supported, and that it is appropriate for the type of endpoint you're configuring.

Reviewer: ```suggestion
    to use is supported for the type of endpoint you're configuring.
```

ach providing a simple, straightforward way to interact with the Gateway.

* ``get_route(name: str) -> Route``: This function allows users to retrieve a specific ``route`` configuration from the `MLflow Gateway` service.

* ``search_routes(search_filter: Optional[str] = None) -> List[Route]``: This function provides the ability to fetch all the routes from the Gateway service. The search filter is not functional currently and the function returns all the routes irrespective of the filter. If an argument is passed, an exception will be thrown to warn of this current lack of functionality.

* ``query(route: str, data)``: This function makes it easy to send a ``query`` to a configured service through a named route on the Gateway Server. It handles the specifics of interfacing with the underlying ``MlflowGatewayClient`` and the configured ``route``, providing a simpler way to issue requests and get responses.

With the ``fluent`` API, working with the `MLflow Gateway` is as easy as making a function call.
It provides a cleaner, more Pythonic way of interacting with the Gateway service, freeing you to
focus on what's really important - getting the results you need from model services.

> To use the ``fluent`` API, here are some examples:

Reviewer 2: Maybe "For the ``fluent`` API, here are some examples:" ? "To use..." doesn't quite parallel "here are..."

          config:
            openai_api_base: https://api.openai.com/v1
            openai_api_key: $OPENAI_API_KEY

      - name: embeddings
        type: llm/v1/embeddings
        model:
          provider: openai
          name: text-embedding-ada-002
          config:
            openai_api_base: https://api.openai.com/v1
            openai_api_key: $OPENAI_API_KEY

Save this file to a location on the system that is going to be running the `MLflow Gateway` server.

Step 4: Start the Gateway Service
---------------------------------
You're now ready to start the Gateway service!

Use the ``mlflow gateway start`` command and specify the path to your configuration file:

.. code-block:: sh

    mlflow gateway start --config-path config.yaml --port {port} --host {host} --workers {worker count}

If you do not specify the host, a localhost address will be used.

If you do not specify the port, port 5000 will be used.

The worker count for gunicorn is defaulted as 2 workers.

Step 5: Access the Interactive API Documentation
------------------------------------------------
The `MLflow Gateway` service provides an interactive API documentation endpoint that you can use to explore

> and test the exposed routes. Navigate to http://{host}:{port}/ (or http://{host}:{port}/docs) in your browser to access it.

Reviewer 2: This renders weird (the http:// renders as an empty link). Not sure how to fix this.

You can now send requests to the exposed routes. Here's an example of how to send a request using Python:

**Note:** Replace "http://localhost:5000/gateway/routes/completions" with the URL of your Gateway service and
the route you want to send a request to. You can replace completions with chat or embeddings to send requests to those routes.

.. code-block:: python

    import requests

    data = {
        "prompt": "Is this a test?",
    }

    response = requests.post(
        "http://localhost:5000/gateway/routes/completions",
        json=data,
    )

    print(response.json())

The response printed to ``stdout`` will be:

.. code-block:: python

    {
        "candidates": [
            {
                "message": {"role": "assistant", "content": "\n\nYes, this is a test."},
                "metadata": {"finish_reason": "stop"},
            }
        ],
        "metadata": {
            "input_tokens": 13,
            "output_tokens": 7,
            "total_tokens": 20,
            "model": "gpt-3.5-turbo-0301",
            "route_type": "llm/v1/chat",
        },
    }

Step 7: Send Requests Using the Fluent API
------------------------------------------

> Here's an example of how to use it to send a chat request using the ``fluent`` API:

Reviewer 2: s/it/the ``fluent`` api/

.. code-block:: python

    from mlflow.gateway import query, set_gateway_uri

    set_gateway_uri(gateway_uri="http://localhost:5000")

    response = query(
        "chat",
        {"messages": [{"role": "user", "content": "What is the best day of the week?"}]},
    )

    print(response)

**Note:** Remember to change the uri definition to the actual uri of your Gateway server.

The returned response will be in this data structure (the actual content and token values will likely be different):

.. code-block:: python

    {
        "candidates": [
            {
                "message": {
                    "role": "assistant",
                    "content": "\n\nIt's hard to say what the best day of the week is.",
                },
                "metadata": {"finish_reason": "stop"},
            }
        ],
        "metadata": {
            "input_tokens": 13,
            "output_tokens": 15,
            "total_tokens": 28,
            "model": "gpt-3.5-turbo-0301",
            "route_type": "llm/v1/chat",
        },
    }

Step 8: Send Requests Using the Client API
------------------------------------------

> Here's an example of how to use it to send a completions request using the ``MlflowGatewayClient`` API:

Reviewer 2: s/it/the ``client`` API/

URL for the API and an environment variable for the API key.

Here is an example of a simple single-route configuration:

.. code-block:: yaml

    routes:
      - name: chat
        type: llm/v1/chat
        model:
          provider: openai
          name: gpt-3.5-turbo
          config:
            openai_api_base: https://api.openai.com/v1
            openai_api_key: $OPENAI_API_KEY

In this example, we define a route named ``chat`` that corresponds to the ``llm/v1/chat`` type, which
will use the ``gpt-3.5-turbo`` model from `OpenAI` to return query responses from the `OpenAI` service.

The beauty of the Gateway Configuration lies in its flexibility and ease of use.
If the configuration file is updated at the target location, the `Mlflow Gateway` service can hot-swap
routes, allowing new providers or model types to be incorporated without requiring changes to
the applications interfacing with the gateway. The hot-swap functionality is autonomous and
requires no server downtime or manual restarting of the service.
This allows for simple, seamless changes to the underlying models and providers while keeping
your applications steady and reliable.

> In order to define an API KEY for a given provider, there are two primary options:

Reviewer 3: There are three, right?

1. In config YAML
2. Environment variable
3. File

Author: good catch. updated the listing :)

1. Directly include it in the YAML configuration file.
2. Use an environment variable to store the API key and reference it in the YAML configuration file.

If you choose to include the API key directly, replace ``$OPENAI_API_KEY`` in the YAML file with your
actual API key.

If you prefer to use an environment variable (recommended), you can define it in your shell
environment. For a Linux system, you can set the environment variable in the ``.bashrc`` file:

1. Open your ``.bashrc`` file in a text editor. For example, using nano:

.. code-block:: bash

     nano ~/.bashrc

2. Scroll down to the end of the file and add the following line:

.. code-block:: bash

     export OPENAI_API_KEY="your_openai_api_key"

Replace "your_openai_api_key" with your actual `OpenAI` API key.

3. Save the file and exit the text editor. If you're using nano, you can do this by pressing
Ctrl+X, then Y to confirm that you want to save the changes, and finally Enter to confirm
the file name.

4. Reload the .bashrc file with the following command:

.. code-block:: bash

>      source ~/.bashrc

Reviewer 3: I don't think we need to mention `nano` or `.bashrc`. This can make users wonder "Does the key need to set in .bashrc?" or "Does the gateway only support bash"?

Reviewer 3: We can just say how to set an API key.

Author: great point. Removing.

* **Request Handling**: When a user sends a request to the `Mlflow Gateway`, it takes the request, adds the necessary authentication (using the API keys it manages), and forwards the request to the correct provider (as specified in the configuration files).

* **Response Forwarding**: The Gateway receives the response from the provider, and then sends this response back to the user.

By centralizing key management, the `MLflow Gateway` dramatically reduces the risk of keys being
lost, misused, or accessed by unauthorized individuals. This approach also simplifies the process
for end users - they no longer need to worry about managing API keys, and can instead focus on
making requests and working with the responses.

Furthermore, this centralized system allows you to change providers or models easily. If you need
to switch models or update API keys, you can do so in a central location without requiring any
changes from your end users. This makes the `MLflow Gateway` a robust and versatile solution for
API key management and service integration within your organization.

> Client API

Reviewer 3: Do we need this section? Can we just add a link to the API doc?
